doc/rfc/5910-consistent-sql-lua-types.md

```diff
@@ -0,0 +1,1687 @@
+    for conversion from `number` as well;
+
+-   Similarly, conversion to `double` might be performed from `string`,
+    `integer`, another `float`, or `number`;
+
+-   `boolean`s can only be converted to `boolean`s;
+
+-   `varbinary` may be converted to `string` and `varbinary`. The reverse is
+    not true for string to varbinary conversion, it's not allowed for
+    implicit cast (but is allowed for _explicit_ cast);
+
+-   Aggregate types like `array` or `map` can only be converted to array or
+    map correspondingly. This is already in the code and does not
+    require any addition after native ARRAY/MAP support will be added to the
+    SQL parser;
+
+
+### Proposed implicit cast table with newer types
+
+Surprisingly, not much change needs to be introduced to the current implicit
+conversation table shown above, for all the new SQL types that we speak about
+in this paper. Array and map are already there and in proper places, and
+omitted there - new scalar uuid type, and several datetime types.
+
+With the addition of newer SQL types (uuid, datetime, etc.) we would
+have the following implicit conversion table:
+
+<img src="media/image18.png" />
+
+-   Date/time/timestamp/intervals entries here are just copy-pasted from
+    the ISO SQL:2016 table above, and may be changed in the final
+    implementation;
+-   If you pay attention to the decimal row then you could recognize
+    that it behaves like other numeric types;
+
+
+## Explicit casts
+
+
+In SQL we could use several ways to explicitly cast from one argument
+type to another, either via `CAST(x AS T)` or via a special purpose
+built-in function.
+
+### Example, ClickHouse explicit casts
+
+ClickHouse is a good example of using both built-in functions and CAST
+expressions for explicit cast of given value to particular target type -
+https://clickhouse.tech/docs/en/sql-reference/functions/type-conversion-functions/
+
+-   `toUInt(8|16|32|64|256)OrZero`
+-   `toUInt(8|16|32|64|256)OrNull`
+-   `toFloat(32|64)`
+-   `toFloat(32|64)OrZero`
+-   `toFloat(32|64)OrNull`
+-   `toDate`
+-   `toDateOrZero`
+-   `toDateOrNull`
+-   `toDateTime`
+-   `toDateTimeOrZero`
+-   `toDateTimeOrNull`
+-   `toDecimal(32|64|128|256)`
+-   `toString`
+-   `toFixedString`
+-   `toInterval(Year|Quarter|Month|Week|Day|Hour|Minute|Second)`
+-   `CAST(x, T)`
+-   `CAST(x AS t)`
+-   And `accurateCast(x, T)` which checks for overflow and raises run-time
+    exceptions;
+
+### Current explicit conversion table
+
+With the exception of Oracle and Microsoft SQL-Server we did not show
+explicit casts tables (for Oracle there was a separate function, and for
+SQL-Server it was integrated to the common table), but the principle
+is simple: use CAST() or function.
+
+Here is the currently implemented table of explicit conversions for
+VDBE:
+
+<img src="media/image19.png" />
+
+### Proposed table for explicit conversions with newer types
+
+<img src="media/image20.png" />
+
```

Comment:
All problems from above table, plus

`interval -> timestamp` - does it mean we set a basic date for all intervals to start with to become a timestamp? Then why `timestamp +> interval` using the same basic date?
`decimal -> boolean` - how? Why not reverse?

Reply:
> `decimal -> boolean` - how? Why not reverse?

Yup, removed all pairs from boolean with exception of string.

> `interval -> timestamp` - does it mean we set a basic date for all intervals to start with to become a timestamp?

This is mostly done to make additions of intervals to timestamp, not assignments. But they do make some sense in any case.